contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,439 @@
+   and deciding if a generated Object corresponds to a particular revision 
+   will require testing each Object for membership in the `OwnerReferences` 
+   of all ControllerRevisions.
+1. Note that, since we are labeling the generated Objects to indicate their
+ providence with respect to the version of the controller's target Object state,
+ we are susceptible to downstream mutations by other controllers changing the 
+ controller's product. The best we can do is guarantee that our product meets 
+ the specification at the time of creation. If a third party mutates the product
+ downstream (as long as it does so in a consistent and intentional way), 
+ we don't want to recall it and make it conform to the original specification. 
+ This would cause the controllers to "fight" indefinitely. 
+   
+At the cost of the complexity of implementing both labeling and ownership, 
+controllers may use a combination of both approaches to mitigate the 
+deficiencies of each.
+
+### Version Equivalence
+When the target Object state of a specification type Object is revised, we wish 
+to minimize the number of mutations to generated Objects as the controller seeks
+to conform the system to its target state. That is, if a generated Object 
+already conforms to the revised target Object state, it is imperative that we 
+do not mutate it. 
+
+Failure to implement this correctly could result in the simultaneous rolling 
+restart of every Pod in every StatefulSet and DaemonSet in the system when 
+additions are made to PodTemplateSpec during a master upgrade. It is therefore 
+necessary to determine if the current target Object state is equivalent to a 
+prior version. 
+
+Since we [track the version of](#version-tracking) of generated Objects, this 
+reduces to deciding if the version of the target Object state associated with 
+the generated Object is equivalent to the current target Object state.
+Even though  [hashing](#hashing) is used to generate the `.Name` of the 
+ControllerRevisions used to encapsulate versions of the target Object state, as 
+we do not require cryptographically strong collision resistance, and given we 
+use a [collision resolution](#collision-resolution) technique, we can't use the 
+[generated names](#unique-name-generation) of ControllerRevisions to decide 
+equality.
+
+We propose that two ControllerRevisions can be considered equal if their 
+`.Data` is equivalent, but that it is not sufficient to compare the serialized 
+representation of the their `.Data`. Consider that the addition of new fields 
+to the Objects that represent the target Object state may cause the serialized 
+representation of those Objects to be unequal even when they are semantically 
+equivalent.
+
+The controller should deserialize the values of the ControllerRevisions 
+representing their target Object state and perform a deep, semantic equality 
+test. Here all differences that do not constitute a mutation to the target 
+Object state are 
```

Comment:
@kow3ns  may be mention that we would do this deep semantic equality only when the hash is same